sherpa/optmethods/transformation.py

```diff
@@ -0,0 +1,260 @@
+#
+#  Copyright (C) 2022
+#  Smithsonian Astrophysical Observatory
+#
+#
+#  This program is free software; you can redistribute it and/or modify
+#  it under the terms of the GNU General Public License as published by
+#  the Free Software Foundation; either version 3 of the License, or
+#  (at your option) any later version.
+#
+#  This program is distributed in the hope that it will be useful,
+#  but WITHOUT ANY WARRANTY; without even the implied warranty of
+#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+#  GNU General Public License for more details.
+#
+#  You should have received a copy of the GNU General Public License along
+#  with this program; if not, write to the Free Software Foundation, Inc.,
+#  51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+#
+
+import numpy
+import inspect
+
+from sherpa.models.parameter import hugeval
+
+__all__ = ('check_transformation', 'NoTransformation', 'Transformation')
+
+class NoTransformation:
+    '''An invariant transformation.  Model evaluations are done in the
+    ext(ernal) coordinate, but within the optimization algorthm calcuations
+    are done in the int(ernal) coordinate system. So final output from the
+    optimization must re-transform from internal to external coordinate.
+    However, since this transformation is invariant the internal and external
+    coordinates are identical.
+    '''
+
+    def __init__(self):
+        self.lo = None
+        self.hi = None
+        return
+
+    def calc_covar(self, m, n, xint, info, fjac):
+        covar = None
+        if info > 0:
+            fjac = numpy.reshape(numpy.ravel(fjac, order='F'),
+                                 (m, n), order='F')
+            if m != n:
+                covar = fjac[:n, :n]
+            else:
+                covar = fjac
+        return covar
+
+    def chain_rule(self, x):
+        return numpy.ones_like(x)
+
+    def ext2int(self, x):
+        return x
+
+    def int2ext(self, x):
+        return x
+
+    def set_bounds(self, lo, hi):
+        self.lo = lo
+        self.hi = hi
+
+
+class Transformation(NoTransformation):
+    ''' See the doc of the NoTransformation base class.
+    This class enables an implementation of an unbounded optimization algorithm
+    into a bounded one. It is capable of transforming each free parameter using
+    the appropiate transformation depending on its boundary conditions of each
+    free parameter.
+    It is recommended that the class should only be used if a free parameter,
+    after a fit, is sufficiently close to upper and/or lower limits. The reason
+    for this is some Sherpa model parameters have large, upper and/or lower
+    limits and one could lose resolution when transforming. A parameter
+    with a low, equal to -hugeval, or high bound,  equal to hugeval,
+    is considered to have only a low or high bound only.
```

Comment:
delete the last 'only' in the line above, so it reads
"is considered to have only a low or high bound.